sqldelight-runtime/src/main/java/com/squareup/sqldelight/Transacter.kt

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (C) 2018 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.sqldelight
+
+import com.squareup.sqldelight.Transacter.Transaction
+import com.squareup.sqldelight.db.SqlDatabase
+
+/**
+ * A transaction-aware [SqlDatabase] wrapper which can begin a [Transaction] on the current thread.
+ */
+abstract class Transacter(
+  private val helper: SqlDatabase,
+  private val transactions: ThreadLocal<Transaction> = ThreadLocal()
+) {
+  /**
+   * For internal use, performs [function] immediately if there is no active [Transaction] on this
+   * thread, otherwise defers [function] to happen on transaction commit.
+   */
+  internal fun deferAction(function: () -> Unit) {
+    val transaction = transactions.get()
+    if (transaction != null) {
+      transaction.postCommitHooks.add(function)
+    } else {
+      function()
+    }
+  }
+
+  /**
+   * Starts a transaction and runs [body] in that transaction.
+   *
+   * @throws IllegalStateException if a [Transaction] is currently active on this thread.
+   * @see transaction
+   */
+  fun newTransaction(body: Transaction.() -> Unit) {
+    if (transactions.get() != null) throw IllegalStateException("Already in a transaction")
+    transaction(body)
+  }
+
+  /**
+   * Starts a [Transaction] and runs [body] in that transaction.
+   */
+  fun transaction(body: Transaction.() -> Unit) {
+    val enclosing = transactions.get()
+    val transaction = Transaction()
+
+    if (enclosing == null) {
+      helper.getConnection().beginTransaction()
+    }
+
+    try {
+      transactions.set(transaction)
+      transaction.body()
+      transaction.successful = true
+    } catch (e: RollbackException) {
+      if (enclosing != null) throw e
+    } finally {
+      transactions.set(enclosing)
+      if (enclosing == null) {
+        if (!transaction.successful || !transaction.childrenSuccessful) {
+          helper.getConnection().rollbackTransaction()
+          while (transaction.postRollbackHooks.isNotEmpty()) {
+            transaction.postRollbackHooks.removeAt(0).invoke()
```

Comment:
Are these user code? I'm a bit worried about the exception implications here.

Reply:
there should still be database integrity since the rollback has already been executed. They can be user code so totally possible an exception happens here. Only ever runs when we are no longer in a transaction though so that should be okay.

Any cases in particular you think might cause problems? I can add some tests to verify behavior when exceptions happen at certain times.

Reply:
I'm mostly thinking about when the block throws some exception and then this throws. In Java bytecode, at least, the original exception will be lost. The way that Rx handles this is that it try/catches the second callback and then if both fail it wraps both exceptions in a `CompositeException` that encapsulates both.

Reply:
yup, makes sense and would definitely happen. I'll add an `@Ignore` test for now with the case and a TODO since the `CompositeException` code looks pretty complex and would probably be better in its own PR.

Reply:
Yeah it's a pain-in-the-ass edge case! 👍